### PR TITLE
refactor(remote): overhaul remote package

### LINKED ISSUE #1449: temp registry server isn't properly storing published references
in v0.9.9 we introduced a regression where qri-io/temp_registry_server is no longer doing the right thing when we publish. Need to investigate & fix. this is blocking tests on both qri and desktop. Desktop e2e tests rely on qri-io/temp_registry_server, and the new preview command needs a temp registry it can accurately publish to for tests

### PR DESCRIPTION
**Pull request by b5**
closes #1449 

### Overhaul Remote package
Ok this is too many changes at once, but we've got lots going on these days, so such is life. Most of these changes stem from "qri add" suddenly ceasing to work, which upon investigation seems like it could stem from a number of problems,
at least some of which derive from blending calls to `CanonicalizeDatasetRef` and `ResolveRef` in the same code path.

overhaul a bunch of remote resolution stuff, including a better mock remote client, dropping `registry.RegClient`'s `ResolveRef` method (it was duplicating code in remote.Client), and adding the `done` interface to `remote.Client`.
Updating the remote package clarifies that ResolveRef really does need to set the peerID as part of reference resolution. Long term this will need to evolve as collaboration necessitates the transition to dataset authors using a shared secret keypair instead of piggybacking on the author's keypair, but either way: reference resolution needs to set a public key verification mechanism so logbook data can be confirmed over the wire. This commit makes a spec change enforcing such a requirement

BREAKING CHANGE:
* all remote hooks pass dsref.Ref instead of reporef.DatasetRef arguments
* ResolveRef spec now requires public key ID be set as part of reference resolution

### Adding a `Bus` method to the `repo` interface

#1430 introduced a number of places where we're using `nil.EventBus`, and left some wiring questions unanswered. To properly cleanup "newTempRepo" we need to make sure it has a real event bus. We're trying to avoid adding things to the repo interface, but there's no way around this one without moving up into instance land, which isn't possible for numerous test suites, because they're below lib.

This change also uncovered a bug where TempRepo struct was retaining the internal repo, which would have crossed wires with the provided context argument. This might help with some of the "repo already open" errors we're seeing in the cmd package tests (I don't expect this is the *only* source of this error, though).

also: remove NewIPFSRepoFromDir. it has zero callers.

### CODE REVIEW

File: lib/lib.go [hunk 1]
```diff
@@ -661,19 +679,29 @@ func (inst *Instance) Connect(ctx context.Context) (err error) {
 	// old instance, we run into issues where the online instance can't "see"
 	// the additions. We fix that by re-initializing the client and remote with the new
 	// instance
-	if inst.remoteClient, err = remote.NewClient(inst.node); err != nil {
+	if inst.remoteClient, err = remote.NewClient(ctx, inst.node); err != nil {
```
Comment (ramfox) [inline lib/lib.go:682]: this old ipfs node instance reference business is so hairy.

Trying to think of ways we can refactor in the future so it doesn't smell so much. Should `remoteClient`, `remote`, & `registry` really be the responsibility of the `node` (which you can argue exists to manage the communications btw this instance and other instances.) Then the `remoteClient` & `remote` update can exist closer to where the actual ipfs node replacement happens.

Should `remoteClient` or `remote` only initialize right before they are used?

File: logbook/logbook.go [hunk 2]
```diff
@@ -790,12 +790,28 @@ func (book *Book) ResolveRef(ctx context.Context, ref *dsref.Ref) (string, error
 	}
 	ref.InitID = initID
 
+	var branchLog *BranchLog
 	if ref.Path == "" {
-		br, err := book.branchLog(ctx, initID)
+		branchLog, err = book.branchLog(ctx, initID)
```
[the remark this answers is not in the document]
Comment (ramfox) [reply logbook/logbook.go:795]: This comment is actually about lines 795 - 814

File: lib/remote.go [hunk 3]
```diff
@@ -75,10 +75,11 @@ func (r *RemoteMethods) Publish(p *PublicationParams, res *dsref.Ref) error {
...
 	ref.ProfileID = pro.ID.String()
 
-	datasetRef := reporef.RefFromDsref(ref)
-	if err = r.inst.RemoteClient().PushDataset(ctx, datasetRef, addr); err != nil {
+	if err = r.inst.RemoteClient().PushDatasetVersion(ctx, ref, addr); err != nil {
 		return err
 	}
+
+	datasetRef := reporef.RefFromDsref(ref)
 	datasetRef.Published = true
```
[the remark this answers is not in the document]
Comment (b5) [reply lib/remote.go:83]: yeah, I think this is part of a follow-up PR that fixes publish properly. I'd like to land this PR and then move to fixing the published flag

File: lib/remote.go [hunk 4]
```diff
@@ -94,16 +95,18 @@ func (r *RemoteMethods) Unpublish(p *PublicationParams, res *dsref.Ref) error {
 		return checkRPCError(r.inst.rpc.Call("RemoteMethods.Unpublish", p, res))
 	}
 
-	ref, err := repo.ParseDatasetRef(p.Ref)
+	ref, err := dsref.ParseHumanFriendly(p.Ref)
 	if err != nil {
+		if err == dsref.ErrNotHumanFriendly {
+			return fmt.Errorf("can only unpublish entire dataset. run unpublish without a reference")
```
[the remark this answers is not in the document]
Comment (b5) [reply lib/remote.go:101]: no this is the right fix for sure

File: remote/browse.go [hunk 5]
```diff
@@ -89,16 +88,16 @@ var _ Previews = (*RepoPreviews)(nil)
...
 // Preview gets a preview for a reference
 func (rp RepoPreviews) Preview(ctx context.Context, _, refStr string) (*dataset.Dataset, error) {
-	ref, err := repo.ParseDatasetRef(refStr)
+	ref, err := dsref.Parse(refStr)
 	if err != nil {
 		return nil, err
 	}
 
-	if err = repo.CanonicalizeDatasetRef(rp.Repo, &ref); err != nil {
+	if _, err := rp.Repo.Logbook().ResolveRef(ctx, &ref); err != nil {
```
[the remark this answers is not in the document]
Comment (b5) [reply remote/browse.go:96]: This is a really good point. Remote should really be using the same pattern we talked about for p2p. I'll make this change: passing a `localResolver` param to the `remote.NewRemote` constructor

**Comment by b5** [comment]
Ok, so I saw @ramfox's [review comment](https://github.com/qri-io/qri/pull/1450#discussion_r459000286) and realized now is the best time to properly align the `remote.Client` interface with [RFC0030](https://github.com/qri-io/rfcs/blob/master/text/0030-replace_publish_clone_with_push_pull.md) on push & pull, so I've spent the day re-working things.

This is basically a new PR, but under the same heading & topic. Not sure if others want to do a full review or just give a quick browse, so I'm re-requesting.

here's what push & pull look like now:

http://www.giphy.com/gifs/Xe9cWplJ3Ygikf6nUl

http://www.giphy.com/gifs/Lqa1O6NSqatR1jRlis

**Pull request merged**